lib/estimator/estimator.rb

```diff
@@ -0,0 +1,89 @@
+module Estimator
+  class Estimate < ActiveRecord::Base
+    require 'active_shipping'
+
+    def self.query(ship_params)
+      @api_call_ok = true
+      ship_array = shipment_array(ship_params)
+      ups_rates = get_rates(ship_array, ups)
+      usps_rates = get_rates(ship_array, usps)
+      quote = assemble_quote(ups_rates, usps_rates)
+      if @api_call_ok
+        return quote
+      else
+        return nil
+      end
+      @api_call_ok = true
+    end
+
+    def self.ups
+      ActiveShipping::UPS.new(login: 'shopifolk', password: 'Shopify_rocks', key: '7CE85DED4C9D07AB')
+    end
+
+    def self.usps
+      ActiveShipping::USPS.new(login: '677JADED7283')
+    end
+
+    def self.shipment_array(ship_params)
+      #from params, goes through packages and gets each one in the format needed to make the call to active shipping
+      #the shipment array is an array of hashes. each hash has a key and the value is an object that can be used by active shipping to make the call to the shipping service
+      #notice that because there are multiple package items, each package needs to be packed by calling on the pack_items method
+      shipment_array = []
+      ship_params[:packages].each do |package|
+        package_info = pack_items(package)
+        active_origin = ActiveShipping::Location.new(package[:origin])
+        active_destination = destination(ship_params)
+        active_package = ActiveShipping::Package.new(package_info[:weight], package_info[:dimensions], :units => :imperial, :value => 10) #had to add value for delivery date estimate
+
+        package_hash = {origin: active_origin,
+                        destination: active_destination,
+                        package: active_package
+                        }
+        shipment_array << package_hash
+      end
+      return shipment_array
+    end
+
+    def self.pack_items(package)
+      #different items from the same merchant will be packed together into one package
+      #a simple way to do this is to find the longest length and and add together the widths and heights. return the total weight and the dimensions of this package.
+      #this method assumes the length is the longest dimension
+      weight = package[:package_items].map {|item| item[:weight].to_i}.sum
```

Comment:
Definitely appreciate the use of comments as a way to list assumptions going into the way the method is written. Thumbs up!